File: src/main/java/com/example/gava/repository/TodoRepository.java
package com.example.gava.repository;

import com.example.gava.entity.Todo;
import org.springframework.data.r2dbc.repository.Query;
import org.springframework.data.repository.reactive.ReactiveCrudRepository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.time.LocalDate;
import java.time.LocalDateTime;

public interface TodoRepository extends ReactiveCrudRepository<Todo, Long> {
    Flux<Todo> findAllByOrderByPriorityAsc();
    Flux<Todo> findAllByOrderByPriorityDesc();
    Flux<Todo> findAllByOrderByDueDateAsc();

    @Query("SELECT * FROM todos WHERE DATE(due_date) = :dueDate")
    Flux<Todo> findByDueDate(LocalDate dueDate);
    Flux<Todo> findByDueDateBetween(LocalDateTime startDate, LocalDateTime endDate);

    @Query("SELECT COUNT(*) FROM todos WHERE completed = true AND due_date BETWEEN :startDate AND :endDate")
    Mono<Long> countCompletedTodosBetween(LocalDateTime startDate, LocalDateTime endDate);
}

File: src/main/kotlin/com/example/gava/domain/todo/repository/TodoRepository.kt
package com.example.gava.domain.todo.repository

import com.example.gava.domain.todo.entity.Todo
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import java.time.LocalDate

interface TodoRepository : JpaRepository<Todo, Long> {

    @Query("""
        SELECT t FROM Todo t
        LEFT JOIN FETCH t._subTodos st
        LEFT JOIN FETCH t._groups tg
        LEFT JOIN FETCH t.icon i
        WHERE t.user.id = :userId
        AND (:date IS NULL OR t.date = :date)
        AND (:groupId IS NULL OR tg.id = :groupId)
        AND (:done IS NULL OR t.isCompleted = :done)
    """)
    fun findAllByUserIdAndDateAndGroupIdAndDone(
        userId: Long,
        date: LocalDate?,
        groupId: Long?,
        done: Boolean?
    ): List<Todo>
}
